test/Microsoft.AspNetCore.SignalR.Redis.Tests/TestConnectionMultiplexer.cs

```diff
@@ -0,0 +1,295 @@
+        }
+
+        public void Dispose()
+        {
+            throw new NotImplementedException();
+        }
+
+        public ProfiledCommandEnumerable FinishProfiling(object forContext, bool allowCleanupSweep = true)
+        {
+            throw new NotImplementedException();
+        }
+
+        public ServerCounters GetCounters()
+        {
+            throw new NotImplementedException();
+        }
+
+        public IDatabase GetDatabase(int db = -1, object asyncState = null)
+        {
+            throw new NotImplementedException();
+        }
+
+        public EndPoint[] GetEndPoints(bool configuredOnly = false)
+        {
+            throw new NotImplementedException();
+        }
+
+        public IServer GetServer(string host, int port, object asyncState = null)
+        {
+            throw new NotImplementedException();
+        }
+
+        public IServer GetServer(string hostAndPort, object asyncState = null)
+        {
+            throw new NotImplementedException();
+        }
+
+        public IServer GetServer(IPAddress host, int port)
+        {
+            throw new NotImplementedException();
+        }
+
+        public IServer GetServer(EndPoint endpoint, object asyncState = null)
+        {
+            throw new NotImplementedException();
+        }
+
+        public string GetStatus()
+        {
+            throw new NotImplementedException();
+        }
+
+        public void GetStatus(TextWriter log)
+        {
+            throw new NotImplementedException();
+        }
+
+        public string GetStormLog()
+        {
+            throw new NotImplementedException();
+        }
+
+        public ISubscriber GetSubscriber(object asyncState = null)
+        {
+            return _subscriber;
+        }
+
+        public int HashSlot(RedisKey key)
+        {
+            throw new NotImplementedException();
+        }
+
+        public long PublishReconfigure(CommandFlags flags = CommandFlags.None)
+        {
+            throw new NotImplementedException();
+        }
+
+        public Task<long> PublishReconfigureAsync(CommandFlags flags = CommandFlags.None)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void RegisterProfiler(IProfiler profiler)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void ResetStormLog()
+        {
+            throw new NotImplementedException();
+        }
+
+        public void Wait(Task task)
+        {
+            throw new NotImplementedException();
+        }
+
+        public T Wait<T>(Task<T> task)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void WaitAll(params Task[] tasks)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    public class TestSubscriber : ISubscriber
+    {
+        private static ConcurrentDictionary<RedisChannel, List<Action<RedisChannel, RedisValue>>> _globalSubscriptions =
```

Comment:
Add a comment why do you need both `_globalSubscriptions` and `_subscriptions`. If I read the code correctly it is to prevent from removing a subscription using a different instance that was used to add it.

Reply:
Will add.
The reason is we are faking a redis server via the `_globalSubscriptions` property. So individual servers need a local copy of their own subscriptions and when publishing need the "server" subscriptions to write to.